cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -0,0 +1,354 @@
+/*
+ * Copyright 2019 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+ *
+ */
+
+package com.linkedin.kafka.cruisecontrol.analyzer.goals;
+
+import com.linkedin.kafka.cruisecontrol.analyzer.OptimizationOptions;
+import com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance;
+import com.linkedin.kafka.cruisecontrol.analyzer.ActionType;
+import com.linkedin.kafka.cruisecontrol.analyzer.AnalyzerUtils;
+import com.linkedin.kafka.cruisecontrol.analyzer.BalancingConstraint;
+import com.linkedin.kafka.cruisecontrol.analyzer.BalancingAction;
+import com.linkedin.kafka.cruisecontrol.common.Statistic;
+import com.linkedin.kafka.cruisecontrol.model.Broker;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModel;
+import com.linkedin.kafka.cruisecontrol.model.ClusterModelStats;
+import com.linkedin.kafka.cruisecontrol.model.Replica;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance.ACCEPT;
+import static com.linkedin.kafka.cruisecontrol.analyzer.ActionAcceptance.REPLICA_REJECT;
+import static com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaDistributionAbstractGoal.ChangeType.*;
+
+/**
+ * SOFT GOAL: Generate leadership movement and leader replica movement proposals to ensure that the number of leader replicas
+ * on each broker is
+ * <ul>
+ * <li>Under: (the average number of leader replicas per broker) * (1 + leader replica count balance percentage)</li>
+ * <li>Above: (the average number of leader replicas per broker) * Math.max(0, 1 - leader replica count balance percentage)</li>
+ * </ul>
+ * Also see: {@link com.linkedin.kafka.cruisecontrol.config.KafkaCruiseControlConfig#LEADER_REPLICA_COUNT_BALANCE_THRESHOLD_CONFIG}
+ * and {@link ReplicaDistributionAbstractGoal#balancePercentageWithMargin(OptimizationOptions, double)}.
+ */
+public class LeaderReplicaDistributionGoal extends ReplicaDistributionAbstractGoal {
+  private static final Logger LOG = LoggerFactory.getLogger(LeaderReplicaDistributionGoal.class);
+
+  /**
+   * Constructor for Leader Replica Distribution Goal.
+   */
+  public LeaderReplicaDistributionGoal() {
+  }
+
+  public LeaderReplicaDistributionGoal(BalancingConstraint balancingConstraint) {
+    this();
```

Comment:
I noticed that we keep the call to `this()`, I don't think we need to call `this()`. -- am I missing something?

Reply:
My understanding is that  parametrized constructor does not call default constructor automatically, but here we need to call default constructor since default constructor will call super class's default constructor and we need to initialize the related fields in [ReplicaDistributionAbstractGoal](https://github.com/linkedin/cruise-control/blob/905b9aee243aab05034579a6a7aeb21aa9ef7e3b/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ReplicaDistributionAbstractGoal.java#L42) and [AbstractGoal](https://github.com/linkedin/cruise-control/blob/905b9aee243aab05034579a6a7aeb21aa9ef7e3b/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/AbstractGoal.java#L54).